resolver-dns/src/main/java/io/netty/resolver/dns/DnsCache.java

```diff
@@ -49,17 +49,19 @@
      * @param additionals the additional records
      * @return the cached entries
      */
-    List<DnsCacheEntry> get(String hostname, DnsRecord[] additionals);
+    List<? extends DnsCacheEntry> get(String hostname, DnsRecord[] additionals);
 
     /**
-     * Cache a resolved address for a given hostname.
+     * Create a new {@link DnsCacheEntry} and cache a resolved address for a given hostname.
      * @param hostname the hostname
      * @param additionals the additional records
      * @param address the resolved address
      * @param originalTtl the TLL as returned by the DNS server
      * @param loop the {@link EventLoop} used to register the TTL timeout
+     * @return The {@link DnsCacheEntry} corresponding to this cache entry.
      */
-    void cache(String hostname, DnsRecord[] additionals, InetAddress address, long originalTtl, EventLoop loop);
+    DnsCacheEntry cache(String hostname, DnsRecord[] additionals, InetAddress address, long originalTtl,
```

Comment:
I debated having a factory but the cache is the one that knows what type of object it will create. Having a factory would mean additional parameters in the resolver, and the cache/factory would have to be coordinated such that the type of object generated by the factory is compatible with the cache. Also if the cache wants to have some implementation details about the type of object this would be leaked by the factory and/or require casting.

Previously a cache object was created outside the cache before this method was called, and then again inside this method (which seems even more leaky). The `DnsCacheEntry` API only exposes "read only" accessors so I don't think we have to worry about external modifications. Someone could of course attempt to cast the object, but the `DefaultDnsCacheEntry` is package private so I don't think that is likely to be a problem.